### PR TITLE
fix(dehoist): question usernames as well

### PR DESCRIPTION
**Pull request by cfanoulis**
Fixes recent issue popping up in Discord

Tested, needs reviews please
Target: weekend release

### CODE REVIEW

File: src/commands/Moderation/dehoist.ts [hunk 1]
```diff
@@ -27,16 +27,18 @@ export default class extends SkyraCommand {
 		const response = await message.sendLocale('SYSTEM_LOADING', []);
 
 		for (const [memberId, memberTag] of members.manageableMembers()) {
-			if (memberTag.nickname && memberTag.nickname.codePointAt(0)! < this.kLowestCode) {
+			const displayName = memberTag.nickname ?? this.client.users.get(memberId)?.username;
+
+			if (displayName && displayName.codePointAt(0)! < this.kLowestCode) {
 				// Replace the first character of the offending user's with a downwards arrow, bringing'em down, down ,down
-				const newNick = `🠷${memberTag.nickname.slice(1)}`;
+				const newNick = `🠷 ${displayName.slice(2)}`;
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: But... why remove two? If anything what you should do is:

1. Continue early if `displayName` is undefined.
1. Store `displayName.codePointAt(0)` as a constant.
1. Check if the constant is lower than `kLowestCode`.
1. If it is, check the byte range, `codePoint <= 0xFF ? 1 : 2`, assign it to a constant.
1. Call `.slice()` with the previous constant.

🎉! UTF-16 support!

**Review by kyranet — CHANGES_REQUESTED** [review]

**Comment by cfanoulis** [comment]
I remove 2 because I use the special char + a space, for visuals

uh I guess I'll do what you did about UTF-18 tho

**Comment by kyranet** [comment]
Most of the symbols used to hoist are "noisy" in a sense, e.g. `!`. For visuals, people add themselves a space. However, this is not always the case. If I set my nickname as `!kyra`, this would change it to `🠷 yra`.

File: src/commands/Moderation/dehoist.ts (hunk 1 above)
Comment (cfanoulis) [reply src/commands/Moderation/dehoist.ts]: resolved in 52b0a266b8949

File: src/commands/Moderation/dehoist.ts [hunk 2]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
 		const response = await message.sendLocale('SYSTEM_LOADING', []);
 
 		for (const [memberId, memberTag] of members.manageableMembers()) {
-			if (memberTag.nickname && memberTag.nickname.codePointAt(0)! < this.kLowestCode) {
+			const displayName = memberTag.nickname ?? this.client.userTags.get(memberId)!.username;
+			if (!displayName) return;
+			const char = displayName.codePointAt(0)!;
+
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
			const displayName = memberTag.nickname ?? this.client.userTags.get(memberId)!.username;
			if (!displayName) continue;

			const char = displayName.codePointAt(0)!;
```

File: src/commands/Moderation/dehoist.ts [hunk 3]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
...
 					await api(this.client)
 						.guilds(message.guild!.id)
 						.members(memberId)
 						.patch({ data: { nick: newNick }, reason: 'Dehoisting' });
 				} catch (error) {
-					errored.push({ oldNick: memberTag.nickname, newNick });
+					errored.push({ oldNick: displayName, newNick });
 				}
 				counter++;
 			}
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: Flatten this with `if (char >= this.kLowestCode) continue;`

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/commands/Moderation/dehoist.ts (hunk 2 above)
Comment (cfanoulis) [reply src/commands/Moderation/dehoist.ts]: resolved in f6c57c8

File: src/commands/Moderation/dehoist.ts (hunk 3 above)
Comment (cfanoulis) [reply src/commands/Moderation/dehoist.ts]: resolved in f6c57c8

File: src/commands/Moderation/dehoist.ts [hunk 4]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
...
-				} catch (error) {
-					errored.push({ oldNick: memberTag.nickname, newNick });
-				}
-				counter++;
+			const displayName = memberTag.nickname ?? this.client.userTags.get(memberId)!.username;
+			if (!displayName) return;
+
+			const char = displayName.codePointAt(0)!;
+
+			if (char > this.kLowestCode) continue;
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
			const char = displayName.codePointAt(0)!;
			if (char > this.kLowestCode) continue;

```

File: src/commands/Moderation/dehoist.ts [hunk 5]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
...
+			// Replace the first character of the offending user's with a downwards arrow, bringing'em down, down ,down
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
			// Replace the first character of the offending user's with an UTF-16 character, bringing'em down, down, down.
```

File: src/commands/Moderation/dehoist.ts [hunk 6]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
...
+			// The ternary cuts 2 characters if the 1st codepoint belongs in UTF-16
+			const newNick = `🠷${displayName.slice(char <= 0xff ? 1 : 2)}`;
+			try {
+				await api(this.client)
+					.guilds(message.guild!.id)
+					.members(memberId)
+					.patch({ data: { nick: newNick }, reason: 'Dehoisting' });
+			} catch (error) {
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
			} catch {
```

File: src/commands/Moderation/dehoist.ts [hunk 7]
```diff
@@ -27,29 +27,34 @@ export default class extends SkyraCommand {
...
+				errored.push({ oldNick: displayName, newNick });
 			}
+			counter++;
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion

			++counter;
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/commands/Moderation/dehoist.ts (hunk 7 above)
Comment (cfanoulis) [reply src/commands/Moderation/dehoist.ts]: does it make a difference? we don't access the value either way, unless one is more performant or something than the other?

Comment (kyranet) [reply src/commands/Moderation/dehoist.ts]: More performant as it doesn't clone the value but just increases it. It's also what we do everywhere else in Skyra.
I also added one line of padding for better readability.

File: src/commands/Moderation/dehoist.ts [hunk 8]
```diff
@@ -27,29 +27,35 @@ export default class extends SkyraCommand {
...
-					await api(this.client)
-						.guilds(message.guild!.id)
-						.members(memberId)
-						.patch({ data: { nick: newNick }, reason: 'Dehoisting' });
-				} catch (error) {
-					errored.push({ oldNick: memberTag.nickname, newNick });
-				}
-				counter++;
+			const displayName = memberTag.nickname ?? this.client.userTags.get(memberId)!.username;
+			if (!displayName) return;
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
			if (!displayName) continue;
```

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by cfanoulis** [comment]
My dumbass forgot and deleted the repo, moved to #1228

**Pull request closed without merging**